weave/trace/weave_client.py

```diff
@@ -961,41 +995,159 @@ def query_costs(
     def _save_object(self, val: Any, name: str, branch: str = "latest") -> ObjectRef:
-        self._save_nested_objects(val, name=name)
+        """Save an object to the weave server and return it's Ref. This is the top
+        level entry point for saving any data to the weave server. Importantly, it
+        will also save all children objects that are "Refable".
+
+        Args:
+            val: The object to save.
+            name: The name to save the object under.
+            branch: The branch to save the object under. Defaults to "latest".
+
+        Returns:
+            An ObjectRef to the saved object.
+        """
+        if isinstance(val, WeaveTable):
+            # TODO: Probably should error here
+            pass
 
-        # typically, this condition would belong inside of the
-        # `_save_nested_objects` switch. However, we don't want to recursively
-        # publish all custom objects. Instead we only want to do this at the
-        # top-most level if requested
-        if get_serializer_for_obj(val) is not None:
-            self._save_and_attach_ref(val)
+        # If it's an Op, use the Op saving logic
+        if isinstance(val, Op):
+            # TODO: Probably should call _save_op directly here (or error)
+            pass
+
+        # Step 1: Recursively save all nested objects
+        self._save_nested_objects(val, name=name)
 
+        # Step 2: Save the object itself
         return self._save_object_basic(val, name, branch)
 
+    @trace_sentry.global_trace_sentry.watch()
+    def _save_nested_objects(self, obj: Any, name: Optional[str] = None) -> Any:
+        """Recursively visits all values, ensuring that any "Refable" objects are
+        saved and reffed.
+        As of this writing, the only "Refable" objects are instances of:
+        - weave.flow.obj.Object
+        - weave.trace.op.Op
+        - weave.trace.Table
+        - weave.trace.vals.WeaveTable
+        This procedure is a bit complicated, so it is worth making the details explicit:
+        1. If the `obj` value already has a `ref`:
+            - If the ref is to the current project, do nothing.
+            - If the ref is to a different project, remove it. (to avoid cross-project references)
+        2. If the `obj` value can be "reffed" (according to one of the above cases), invoke
+            the appropriate "save" function for that type of object, and attach the ref result to `obj`.
+            - `_save_object_basic` (for `weave.flow.obj.Object` instances)
+            - `_save_op` (for `weave.trace.op.Op` instances)
+            - `_save_table` (for `weave.trace.Table` and `weave.trace.vals.WeaveTable` instances)
+        3. Otherwise, traverse all values within `obj` recursively, applying the above logic to each value.
+        Important notes to developers: This method does not return anything - it _mutates_ the
+        values that it traverses (specifically, it attaches `ref` values to them)
```

Comment:
this attaches refs, and also importantly fires off create queries to the server? Might want to include something about how this function actually incurs server latency

Reply:
added in next commit